database/bigquery.js

```diff
@@ -0,0 +1,87 @@
+const fs = require('fs');
+const EventEmitter = require("events");
+const {BigQuery} = require('@google-cloud/bigquery');
+
+const isProduction = process.env.NODE_ENV && process.env.NODE_ENV === 'production';
+
+class RecordBuffer extends EventEmitter {
+    constructor(interval = 5 * 60 * 1000, bufferSize = 100) {
+        super();
+
+        this.interval = interval;
+        this.nextFlush = setTimeout(this.flush.bind(this), this.interval);
+
+        this.bufferSize = bufferSize;
+        this.fileCount = 0;
+        this.bufferedItems = 0;
+        this.currentFile = fs.createWriteStream('bigquery-' + this.fileCount);
+    }
+    put(record) {
+        this.currentFile.write(JSON.stringify(record) + '\n');
+        this.bufferedItems++;
+        if (this.bufferedItems >= this.bufferSize) {
+            this.flush();
+        }
+    }
+    flush() {
+        clearInterval(this.nextFlush);
+        this.nextFlush = setTimeout(this.flush.bind(this), this.interval);
+        if (this.bufferedItems === 0) {
+            return;
+        }
+        this.fileCount++;
+        const currentFile = this.currentFile;
+        this.currentFile = fs.createWriteStream('bigquery-' + this.fileCount);
+        this.bufferedItems = 0;
+        currentFile.end();
+        currentFile.on('finish', () => {
+            this.emit('flush', currentFile.path);
+        });
+    }
+
+}
+
+module.exports = function(config) {
+    if (config.gcp) {
+    } else {
+        console.warn('No GCP/Bigquery configuration present. Skipping Bigquery database.')
+        return;
+    }
+
+    const bigquery = new BigQuery();
+    const recordBuffer = new RecordBuffer();
+    recordBuffer.on('flush', (filename) => {
+        bigquery
+            .dataset(config.gcp.dataset)
+            .table(config.gcp.table)
+            .load(filename, {format: 'JSON'})
+        .then(() => {
+            if (isProduction) {
+                fs.unlink(filename, () => {});
+            }
+        })
+        .catch(e => console.error('error loading into bigquery', e));
+    });
+    return {
+        put: function(pageUrl, clientId, connectionId, clientFeatures, connectionFeatures, streamFeatures) {
+            const d = new Date().getTime();
+            const item = {
+                Date: d - (d % (86400 * 1000)), // just the UTC day
+                DateTime: d,
+                ClientId: clientId,
+                ConnectionId: clientId + '_' + connectionId,
+                PageUrl: pageUrl,
+            };
+
+            Object.assign(item, clientFeatures, connectionFeatures, streamFeatures);
+            if (config.gcp.fields.length) {
+                Object.keys(item).forEach(key => {
+                    if (!config.gcp.fields.includes(key.toLowerCase())) {
```

Comment:
I guess we're expecting `config.gcp.fields` to be a set of valid fields to prevent any issue of including non-valid ones? Should not that definition be part of this module?

Reply:
yes, that is going to be added as part of adding the config. The assumption is that this module will only be called when configured properly.